File: leet/remove-nth-node-from-end-of-list.ts
/**
 * https://leetcode.com/problems/remove-nth-node-from-end-of-list/
 */

export function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
  let dummyHead = new ListNode(0, head);
  let right: any = dummyHead;
  let left: any = dummyHead;

  for (let i = 1; i <= n; i++) {
    right = right.next;
  }

  while (right.next) {
    left = left.next;
    right = right.next;
  }

  left.next = left.next.next;

  return dummyHead.next;
}

class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

export function toList(elements: number[]) {
  if (elements.length === 0) return null;

  let pointer: ListNode | null = null;
  for (let i = elements.length - 1; i >= 0; i--) {
    const node = new ListNode(elements[i]);
    node.next = pointer;
    pointer = node;
  }

  return pointer;
}

export function toArray(list: ListNode | null) {
  if (!list) return [];

  const result: number[] = [];
  let pointer: ListNode | null = list;
  while (pointer) {
    result.push(pointer.val);
    pointer = pointer.next;
  }

  return result;
}

File: returned/python/remove_nth_node_from_end_of_list.py
# https://leetcode.com/problems/remove-nth-node-from-end-of-list/
from typing import Optional, List

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def to_nodes(elements: List[int]) -> Optional[ListNode]:
    if len(elements) == 0:
        return None
    prev = None
    for el in reversed(elements):
        node = ListNode(el)
        node.next = prev
        prev = node
    return prev
    
def to_array(root: Optional[ListNode]) -> List[int]:
    if not root:
        return []
    result = []
    while root:
        result.append(root.val)
        root = root.next
    return result

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)

        pointer = dummy
        counter = 1
        trail = None
        while pointer.next != None:
            if counter == n:
                trail = dummy
            elif counter > n:
                trail = trail.next
            counter += 1
            pointer = pointer.next
            
        trail.next = trail.next.next

        return dummy.next